File: internal/gui/frontend/src/api/services/system.ts
import { SystemService } from '../generated/api.gen';
import type { OpenURLArgs } from '../generated/api.gen';
import { getAPIBaseUrl, defaultFetch } from '../client';

// Singleton instance
let systemServiceInstance: SystemService | null = null;

/**
 * Get or create the system service instance
 */
async function getSystemService(): Promise<SystemService> {
    if (!systemServiceInstance) {
        const baseUrl = await getAPIBaseUrl();
        systemServiceInstance = new SystemService(baseUrl, defaultFetch);
    }
    return systemServiceInstance;
}

/**
 * GetSystemInfo - Drop-in replacement for Wails method
 * Maintains exact same signature as the Wails version
 */
export async function GetSystemInfo(): Promise<Record<string, string>> {
    const service = await getSystemService();
    
    try {
        const response = await service.getSystemInfo();
        return {
            os: response.info.os,
            arch: response.info.arch
        };
    } catch (error) {
        console.error('GetSystemInfo error:', error);
        throw error;
    }
}

/**
 * GetVersion - Drop-in replacement for Wails method
 * Maintains exact same signature as the Wails version
 */
export async function GetVersion(): Promise<any> {
    const service = await getSystemService();
    
    try {
        const response = await service.getVersion();
        return {
            version: response.info.version
        };
    } catch (error) {
        console.error('GetVersion error:', error);
        throw error;
    }
}

/**
 * CheckForUpdate - Check if a newer version is available
 */
export async function CheckForUpdate(): Promise<{ updateAvailable: boolean }> {
    const service = await getSystemService();
    
    try {
        const response = await service.checkForUpdate();
        return {
            updateAvailable: response.updateAvailable
        };
    } catch (error) {
        console.error('CheckForUpdate error:', error);
        throw error;
    }
}

/**
 * OpenURL - Drop-in replacement for Wails method
 * Opens a URL in the user's default browser
 * Maintains exact same signature as the Wails version
 */
export async function OpenURL(url: string): Promise<void> {
    const service = await getSystemService();

    try {
        await service.openURL({ url });
    } catch (error) {
        console.error('OpenURL error:', error);
        throw error;
    }
}

/**
 * ShowWarning - Display a warning dialog to the user
 * Returns true if the user acknowledged the warning
 */
export async function ShowWarning(title: string, message: string): Promise<boolean> {
    const service = await getSystemService();

    try {
        const response = await service.showWarning({ title, message });
        return response.acknowledged;
    } catch (error) {
        console.error('ShowWarning error:', error);
        throw error;
    }
}

/**
 * NvidiaGPUInfo represents NVIDIA GPU information
 */
export interface NvidiaGPUInfo {
    available: boolean;  // Whether an NVIDIA GPU with drivers is detected
    name: string;        // GPU name (e.g., "NVIDIA GeForce RTX 3080"), empty if not available
    vramMiB: number;     // Total VRAM in MiB, 0 if not available
}

/**
 * GetNvidiaGPUInfo - Get NVIDIA GPU information
 * Returns availability, GPU name, and VRAM in MiB
 */
export async function GetNvidiaGPUInfo(): Promise<NvidiaGPUInfo> {
    const service = await getSystemService();

    try {
        const response = await service.getNvidiaGPUInfo();
        return {
            available: response.info.available,
            name: response.info.name,
            vramMiB: Number(response.info.vramMiB)
        };
    } catch (error) {
        console.error('GetNvidiaGPUInfo error:', error);
        // Return unavailable on error
        return {
            available: false,
            name: '',
            vramMiB: 0
        };
    }
}

File: internal/api/services/system.go
package services

import (
	"context"
	"net/http"
	goruntime "runtime"

	"github.com/rs/zerolog"

	"github.com/tassa-yoniso-manasi-karoto/langkit/internal/api"
	"github.com/tassa-yoniso-manasi-karoto/langkit/internal/api/generated"
	"github.com/tassa-yoniso-manasi-karoto/langkit/internal/executils"
	"github.com/tassa-yoniso-manasi-karoto/langkit/internal/ui"
	"github.com/tassa-yoniso-manasi-karoto/langkit/internal/ui/dialogs"
	"github.com/tassa-yoniso-manasi-karoto/langkit/internal/version"
)

// Compile-time check that SystemService implements api.Service
var _ api.Service = (*SystemService)(nil)

// SystemService implements the WebRPC SystemService interface
type SystemService struct {
	logger  zerolog.Logger
	handler http.Handler
}

// NewSystemService creates a new system service instance
func NewSystemService(logger zerolog.Logger) *SystemService {
	svc := &SystemService{
		logger: logger,
	}
	
	// Create the WebRPC handler
	svc.handler = generated.NewSystemServiceServer(svc)
	
	return svc
}

// Name implements api.Service
func (s *SystemService) Name() string {
	return "SystemService"
}

// Handler implements api.Service
func (s *SystemService) Handler() http.Handler {
	return s.handler
}

// Description implements api.Service
func (s *SystemService) Description() string {
	return "System information and version service"
}

// GetSystemInfo returns the user's operating system and architecture
func (s *SystemService) GetSystemInfo(ctx context.Context) (*generated.SystemInfo, error) {
	return &generated.SystemInfo{
		Os:   goruntime.GOOS,
		Arch: goruntime.GOARCH,
	}, nil
}

// GetVersion returns version information
func (s *SystemService) GetVersion(ctx context.Context) (*generated.VersionInfo, error) {
	return &generated.VersionInfo{
		Version: version.Version,
	}, nil
}

// CheckForUpdate checks if a newer version is available
func (s *SystemService) CheckForUpdate(ctx context.Context) (bool, error) {
	info := version.GetInfo(true)  // Wait for update check
	return info.NewerVersionAvailable, nil
}

// OpenURL opens a URL in the user's default browser
func (s *SystemService) OpenURL(ctx context.Context, url string) error {
	return ui.GetURLOpener().OpenURL(url)
}

// ShowWarning displays a warning dialog to the user
func (s *SystemService) ShowWarning(ctx context.Context, title string, message string) (bool, error) {
	s.logger.Debug().Str("title", title).Msg("Showing warning dialog")
	acknowledged, err := ui.GetMessageDialog().ShowMessage(title, message, dialogs.MessageWarning)
	if err != nil {
		s.logger.Error().Err(err).Msg("Failed to show warning dialog")
		return false, err
	}
	return acknowledged, nil
}

// GetNvidiaGPUInfo returns NVIDIA GPU information (availability, name, VRAM)
func (s *SystemService) GetNvidiaGPUInfo(ctx context.Context) (*generated.NvidiaGPUInfo, error) {
	vramMiB := executils.GetNvidiaVRAMMiB()
	gpuName := executils.GetNvidiaGPUName()

	return &generated.NvidiaGPUInfo{
		Available: vramMiB > 0 && gpuName != "",
		Name:      gpuName,
		VramMiB:   int64(vramMiB),
	}, nil
}